cou/apps/base.py

```diff
@@ -621,22 +637,41 @@ def _need_current_channel_refresh(self, target: OpenStackRelease) -> bool:
         :return: True if needs to refresh, False otherwise
         :rtype: bool
         """
-        return bool(self.can_upgrade_to) and self.channel_codename <= target
+        return bool(self.can_upgrade_to) and self.current_channel_os_release <= target
 
     def _get_upgrade_charm_step(self, target: OpenStackRelease) -> UpgradeStep:
         """Get step for upgrading the charm.
 
         :param target: OpenStack release as target to upgrade.
         :type target: OpenStackRelease
+        :raises ApplicationError: When the current channel is ahead from expected and the target.
         :return: Step for upgrading the charm.
         :rtype: UpgradeStep
         """
-        if self.channel != self.target_channel(target):
+        channel = self.expected_current_channel(target) if self.need_crossgrade else self.channel
+
+        if (
+            self.current_channel_os_release <= self.current_os_release
+            and channel != self.target_channel(target)
+        ):
             return UpgradeStep(
-                description=f"Upgrade '{self.name}' to the new channel: "
+                description=f"Upgrade '{self.name}' from '{channel}' to the new channel: "
                 f"'{self.target_channel(target)}'",
                 coro=self.model.upgrade_charm(self.name, self.target_channel(target)),
             )
+
+        if (
+            self.current_channel_os_release > self.current_os_release
+            and channel != self.target_channel(target)
+        ):
+            raise ApplicationError(
+                f"'{self.name}' has the channel ahead from expected. The channel '{self.channel}' "
+                f"doesn't match with the expected '{self.expected_current_channel(target)}' or "
+                f"with the target channel '{self.target_channel(target)}'. "
+                "Manual intervention is required."
```

Comment:
```suggestion
                f"The '{self.name}' charm is using channel '{self.channel}'. Channels supported during this transition:['{self.expected_current_channel(target)}', '{self.target_channel(target)}']. Manual intervention is required."
```
w/some newlines where black wants to put them :)